File: src/data/places.ts
import type { Place } from '@/types'

/** Default map centre — Kuala Lumpur city centre. */
export const CITY_CENTER = { lat: 3.1478, lng: 101.6953 }
export const CITY_NAME = 'Kuala Lumpur'

/**
 * Offline place index. Search works with no network at all; when the device is
 * online the geocoder in `services/geocode.ts` layers live results on top.
 */
export const PLACES: Place[] = [
  { id: 'p_klcc', name: 'Suria KLCC', address: 'Jalan Ampang, Kuala Lumpur City Centre', coord: { lat: 3.1578, lng: 101.7123 }, category: 'mall' },
  { id: 'p_twin', name: 'Petronas Twin Towers', address: 'Kuala Lumpur City Centre', coord: { lat: 3.1579, lng: 101.7116 }, category: 'landmark' },
  { id: 'p_klia', name: 'KLIA Terminal 1', address: 'Kuala Lumpur International Airport, Sepang', coord: { lat: 2.7456, lng: 101.7099 }, category: 'airport' },
  { id: 'p_klia2', name: 'KLIA Terminal 2', address: 'KLIA2, Sepang', coord: { lat: 2.7426, lng: 101.6862 }, category: 'airport' },
  { id: 'p_subang', name: 'Sultan Abdul Aziz Shah Airport', address: 'Subang, Selangor', coord: { lat: 3.1304, lng: 101.5493 }, category: 'airport' },
  { id: 'p_sentral', name: 'KL Sentral', address: 'Jalan Stesen Sentral, Brickfields', coord: { lat: 3.1338, lng: 101.6869 }, category: 'transit' },
  { id: 'p_pavilion', name: 'Pavilion Kuala Lumpur', address: 'Jalan Bukit Bintang, Bukit Bintang', coord: { lat: 3.1490, lng: 101.7132 }, category: 'mall' },
  { id: 'p_bb', name: 'Bukit Bintang', address: 'Bukit Bintang, Kuala Lumpur', coord: { lat: 3.1466, lng: 101.7113 }, category: 'area' },
  { id: 'p_midvalley', name: 'Mid Valley Megamall', address: 'Lingkaran Syed Putra, Mid Valley City', coord: { lat: 3.1177, lng: 101.6771 }, category: 'mall' },
  { id: 'p_gardens', name: 'The Gardens Mall', address: 'Mid Valley City, Lingkaran Syed Putra', coord: { lat: 3.1183, lng: 101.6764 }, category: 'mall' },
  { id: 'p_bangsar', name: 'Bangsar Village', address: 'Jalan Telawi 1, Bangsar', coord: { lat: 3.1300, lng: 101.6700 }, category: 'mall' },
  { id: 'p_mont', name: 'Mont Kiara', address: 'Mont Kiara, Kuala Lumpur', coord: { lat: 3.1710, lng: 101.6500 }, category: 'area' },
  { id: 'p_ttdi', name: 'Taman Tun Dr Ismail', address: 'TTDI, Kuala Lumpur', coord: { lat: 3.1450, lng: 101.6300 }, category: 'area' },
  { id: 'p_damansara', name: 'The Curve, Mutiara Damansara', address: 'Jalan PJU 7/3, Petaling Jaya', coord: { lat: 3.1580, lng: 101.6110 }, category: 'mall' },
  { id: 'p_1u', name: '1 Utama Shopping Centre', address: 'Bandar Utama, Petaling Jaya', coord: { lat: 3.1502, lng: 101.6155 }, category: 'mall' },
  { id: 'p_sunway', name: 'Sunway Pyramid', address: 'Bandar Sunway, Subang Jaya', coord: { lat: 3.0726, lng: 101.6069 }, category: 'mall' },
  { id: 'p_pj', name: 'Petaling Jaya SS2', address: 'SS2, Petaling Jaya, Selangor', coord: { lat: 3.1180, lng: 101.6230 }, category: 'area' },
  { id: 'p_shah', name: 'Shah Alam i-City', address: 'Seksyen 7, Shah Alam', coord: { lat: 3.0640, lng: 101.4880 }, category: 'landmark' },
  { id: 'p_puchong', name: 'IOI Mall Puchong', address: 'Bandar Puchong Jaya, Puchong', coord: { lat: 3.0450, lng: 101.6180 }, category: 'mall' },
  { id: 'p_cheras', name: 'Cheras Leisure Mall', address: 'Taman Segar, Cheras', coord: { lat: 3.0930, lng: 101.7420 }, category: 'mall' },
  { id: 'p_ampang', name: 'Ampang Park', address: 'Jalan Ampang, Kuala Lumpur', coord: { lat: 3.1600, lng: 101.7180 }, category: 'transit' },
  { id: 'p_setapak', name: 'Wangsa Walk Mall', address: 'Wangsa Maju, Kuala Lumpur', coord: { lat: 3.2020, lng: 101.7350 }, category: 'mall' },
  { id: 'p_putrajaya', name: 'Putrajaya Sentral', address: 'Presint 7, Putrajaya', coord: { lat: 2.9310, lng: 101.6710 }, category: 'transit' },
  { id: 'p_cyber', name: 'Cyberjaya City Centre', address: 'Cyberjaya, Selangor', coord: { lat: 2.9210, lng: 101.6560 }, category: 'area' },
  { id: 'p_batu', name: 'Batu Caves', address: 'Gombak, Selangor', coord: { lat: 3.2379, lng: 101.6840 }, category: 'landmark' },
  { id: 'p_merdeka', name: 'Merdeka 118', address: 'Jalan Hang Jebat, Kuala Lumpur', coord: { lat: 3.1416, lng: 101.7009 }, category: 'landmark' },
  { id: 'p_masjid', name: 'Masjid Jamek', address: 'Jalan Tun Perak, Kuala Lumpur', coord: { lat: 3.1490, lng: 101.6960 }, category: 'transit' },
  { id: 'p_chinatown', name: 'Petaling Street', address: 'Chinatown, Kuala Lumpur', coord: { lat: 3.1436, lng: 101.6980 }, category: 'landmark' },
  { id: 'p_titiwangsa', name: 'Titiwangsa Lake Gardens', address: 'Titiwangsa, Kuala Lumpur', coord: { lat: 3.1780, lng: 101.7020 }, category: 'landmark' },
  { id: 'p_desapark', name: 'Desa ParkCity', address: 'Desa ParkCity, Kuala Lumpur', coord: { lat: 3.1870, lng: 101.6300 }, category: 'area' },
  { id: 'p_hospital', name: 'Hospital Kuala Lumpur', address: 'Jalan Pahang, Kuala Lumpur', coord: { lat: 3.1730, lng: 101.6990 }, category: 'landmark' },
  { id: 'p_um', name: 'Universiti Malaya', address: 'Jalan Universiti, Kuala Lumpur', coord: { lat: 3.1209, lng: 101.6538 }, category: 'landmark' },
  { id: 'p_ioicity', name: 'IOI City Mall', address: 'Putrajaya, Selangor', coord: { lat: 2.9690, lng: 101.7130 }, category: 'mall' },
  { id: 'p_klgateway', name: 'KL Gateway Mall', address: 'Kerinchi, Kuala Lumpur', coord: { lat: 3.1140, lng: 101.6640 }, category: 'mall' },
  { id: 'p_tbs', name: 'Terminal Bersepadu Selatan', address: 'Bandar Tasik Selatan, Kuala Lumpur', coord: { lat: 3.0760, lng: 101.7110 }, category: 'transit' },
  { id: 'p_bukitjalil', name: 'Bukit Jalil National Stadium', address: 'Bukit Jalil, Kuala Lumpur', coord: { lat: 3.0545, lng: 101.6913 }, category: 'landmark' },
  { id: 'p_pavilion2', name: 'Pavilion Bukit Jalil', address: 'Bukit Jalil, Kuala Lumpur', coord: { lat: 3.0590, lng: 101.6870 }, category: 'mall' },
  { id: 'p_setiawalk', name: 'SetiaWalk Puchong', address: 'Pusat Bandar Puchong', coord: { lat: 3.0230, lng: 101.6180 }, category: 'mall' },
  { id: 'p_klang', name: 'Klang Town Centre', address: 'Klang, Selangor', coord: { lat: 3.0450, lng: 101.4450 }, category: 'area' },
  { id: 'p_genting', name: 'Genting Highlands', address: 'Bentong, Pahang', coord: { lat: 3.4230, lng: 101.7930 }, category: 'landmark' },
]

/** Intercity destinations offered in the Intercity vertical. */
export const INTERCITY_PLACES: Place[] = [
  { id: 'ic_ipoh', name: 'Ipoh', address: 'Perak', coord: { lat: 4.5975, lng: 101.0901 }, category: 'area' },
  { id: 'ic_penang', name: 'George Town, Penang', address: 'Pulau Pinang', coord: { lat: 5.4141, lng: 100.3288 }, category: 'area' },
  { id: 'ic_melaka', name: 'Melaka', address: 'Melaka', coord: { lat: 2.1896, lng: 102.2501 }, category: 'area' },
  { id: 'ic_jb', name: 'Johor Bahru', address: 'Johor', coord: { lat: 1.4927, lng: 103.7414 }, category: 'area' },
  { id: 'ic_seremban', name: 'Seremban', address: 'Negeri Sembilan', coord: { lat: 2.7297, lng: 101.9381 }, category: 'area' },
  { id: 'ic_kuantan', name: 'Kuantan', address: 'Pahang', coord: { lat: 3.8077, lng: 103.3260 }, category: 'area' },
  { id: 'ic_cameron', name: 'Cameron Highlands', address: 'Pahang', coord: { lat: 4.4711, lng: 101.3776 }, category: 'area' },
]

export const ALL_PLACES = [...PLACES, ...INTERCITY_PLACES]

/** Street names used to synthesise "dropped pin" addresses. */
export const STREETS = [
  'Jalan Ampang', 'Jalan Sultan Ismail', 'Jalan Raja Chulan', 'Jalan Tun Razak',
  'Jalan Bukit Bintang', 'Jalan Imbi', 'Jalan Pudu', 'Jalan Kuching',
  'Jalan Ipoh', 'Jalan Cheras', 'Jalan Klang Lama', 'Jalan Duta',
  'Persiaran KLCC', 'Lebuh Ampang', 'Jalan Sultan Hishamuddin', 'Jalan Damansara',
]

export function fuzzySearch(query: string, pool: Place[] = ALL_PLACES, limit = 8): Place[] {
  const q = query.trim().toLowerCase()
  if (!q) return []
  const scored = pool
    .map((place) => {
      const name = place.name.toLowerCase()
      const address = place.address.toLowerCase()
      let score = 0
      if (name.startsWith(q)) score += 100
      else if (name.includes(q)) score += 60
      if (address.includes(q)) score += 25
      // Match on individual words so "mid valley" finds "Mid Valley Megamall".
      const words = q.split(/\s+/)
      if (words.length > 1 && words.every((w) => name.includes(w) || address.includes(w))) score += 40
      return { place, score }
    })
    .filter((s) => s.score > 0)
    .sort((a, b) => b.score - a.score)
    .slice(0, limit)
  return scored.map((s) => s.place)
}

File: lib/data/places.dart
import 'package:latlong2/latlong.dart';

import '../models/models.dart';

/// Default map centre — Kuala Lumpur city centre.
const cityCenter = LatLng(3.1478, 101.6953);
const cityName = 'Kuala Lumpur';

Place _p(String id, String name, String address, double lat, double lng, PlaceCategory c) =>
    Place(id: id, name: name, address: address, coord: LatLng(lat, lng), category: c);

/// Offline place index. Search works with no network at all.
final List<Place> places = [
  _p('p_klcc', 'Suria KLCC', 'Jalan Ampang, Kuala Lumpur City Centre', 3.1578, 101.7123, PlaceCategory.mall),
  _p('p_twin', 'Petronas Twin Towers', 'Kuala Lumpur City Centre', 3.1579, 101.7116, PlaceCategory.landmark),
  _p('p_klia', 'KLIA Terminal 1', 'Kuala Lumpur International Airport, Sepang', 2.7456, 101.7099, PlaceCategory.airport),
  _p('p_klia2', 'KLIA Terminal 2', 'KLIA2, Sepang', 2.7426, 101.6862, PlaceCategory.airport),
  _p('p_subang', 'Sultan Abdul Aziz Shah Airport', 'Subang, Selangor', 3.1304, 101.5493, PlaceCategory.airport),
  _p('p_sentral', 'KL Sentral', 'Jalan Stesen Sentral, Brickfields', 3.1338, 101.6869, PlaceCategory.transit),
  _p('p_pavilion', 'Pavilion Kuala Lumpur', 'Jalan Bukit Bintang, Bukit Bintang', 3.1490, 101.7132, PlaceCategory.mall),
  _p('p_bb', 'Bukit Bintang', 'Bukit Bintang, Kuala Lumpur', 3.1466, 101.7113, PlaceCategory.area),
  _p('p_midvalley', 'Mid Valley Megamall', 'Lingkaran Syed Putra, Mid Valley City', 3.1177, 101.6771, PlaceCategory.mall),
  _p('p_gardens', 'The Gardens Mall', 'Mid Valley City, Lingkaran Syed Putra', 3.1183, 101.6764, PlaceCategory.mall),
  _p('p_bangsar', 'Bangsar Village', 'Jalan Telawi 1, Bangsar', 3.1300, 101.6700, PlaceCategory.mall),
  _p('p_mont', 'Mont Kiara', 'Mont Kiara, Kuala Lumpur', 3.1710, 101.6500, PlaceCategory.area),
  _p('p_ttdi', 'Taman Tun Dr Ismail', 'TTDI, Kuala Lumpur', 3.1450, 101.6300, PlaceCategory.area),
  _p('p_damansara', 'The Curve, Mutiara Damansara', 'Jalan PJU 7/3, Petaling Jaya', 3.1580, 101.6110, PlaceCategory.mall),
  _p('p_1u', '1 Utama Shopping Centre', 'Bandar Utama, Petaling Jaya', 3.1502, 101.6155, PlaceCategory.mall),
  _p('p_sunway', 'Sunway Pyramid', 'Bandar Sunway, Subang Jaya', 3.0726, 101.6069, PlaceCategory.mall),
  _p('p_pj', 'Petaling Jaya SS2', 'SS2, Petaling Jaya, Selangor', 3.1180, 101.6230, PlaceCategory.area),
  _p('p_shah', 'Shah Alam i-City', 'Seksyen 7, Shah Alam', 3.0640, 101.4880, PlaceCategory.landmark),
  _p('p_puchong', 'IOI Mall Puchong', 'Bandar Puchong Jaya, Puchong', 3.0450, 101.6180, PlaceCategory.mall),
  _p('p_cheras', 'Cheras Leisure Mall', 'Taman Segar, Cheras', 3.0930, 101.7420, PlaceCategory.mall),
  _p('p_ampang', 'Ampang Park', 'Jalan Ampang, Kuala Lumpur', 3.1600, 101.7180, PlaceCategory.transit),
  _p('p_setapak', 'Wangsa Walk Mall', 'Wangsa Maju, Kuala Lumpur', 3.2020, 101.7350, PlaceCategory.mall),
  _p('p_putrajaya', 'Putrajaya Sentral', 'Presint 7, Putrajaya', 2.9310, 101.6710, PlaceCategory.transit),
  _p('p_cyber', 'Cyberjaya City Centre', 'Cyberjaya, Selangor', 2.9210, 101.6560, PlaceCategory.area),
  _p('p_batu', 'Batu Caves', 'Gombak, Selangor', 3.2379, 101.6840, PlaceCategory.landmark),
  _p('p_merdeka', 'Merdeka 118', 'Jalan Hang Jebat, Kuala Lumpur', 3.1416, 101.7009, PlaceCategory.landmark),
  _p('p_masjid', 'Masjid Jamek', 'Jalan Tun Perak, Kuala Lumpur', 3.1490, 101.6960, PlaceCategory.transit),
  _p('p_chinatown', 'Petaling Street', 'Chinatown, Kuala Lumpur', 3.1436, 101.6980, PlaceCategory.landmark),
  _p('p_titiwangsa', 'Titiwangsa Lake Gardens', 'Titiwangsa, Kuala Lumpur', 3.1780, 101.7020, PlaceCategory.landmark),
  _p('p_desapark', 'Desa ParkCity', 'Desa ParkCity, Kuala Lumpur', 3.1870, 101.6300, PlaceCategory.area),
  _p('p_hospital', 'Hospital Kuala Lumpur', 'Jalan Pahang, Kuala Lumpur', 3.1730, 101.6990, PlaceCategory.landmark),
  _p('p_um', 'Universiti Malaya', 'Jalan Universiti, Kuala Lumpur', 3.1209, 101.6538, PlaceCategory.landmark),
  _p('p_ioicity', 'IOI City Mall', 'Putrajaya, Selangor', 2.9690, 101.7130, PlaceCategory.mall),
  _p('p_klgateway', 'KL Gateway Mall', 'Kerinchi, Kuala Lumpur', 3.1140, 101.6640, PlaceCategory.mall),
  _p('p_tbs', 'Terminal Bersepadu Selatan', 'Bandar Tasik Selatan, Kuala Lumpur', 3.0760, 101.7110, PlaceCategory.transit),
  _p('p_bukitjalil', 'Bukit Jalil National Stadium', 'Bukit Jalil, Kuala Lumpur', 3.0545, 101.6913, PlaceCategory.landmark),
  _p('p_pavilion2', 'Pavilion Bukit Jalil', 'Bukit Jalil, Kuala Lumpur', 3.0590, 101.6870, PlaceCategory.mall),
  _p('p_setiawalk', 'SetiaWalk Puchong', 'Pusat Bandar Puchong', 3.0230, 101.6180, PlaceCategory.mall),
  _p('p_klang', 'Klang Town Centre', 'Klang, Selangor', 3.0450, 101.4450, PlaceCategory.area),
  _p('p_genting', 'Genting Highlands', 'Bentong, Pahang', 3.4230, 101.7930, PlaceCategory.landmark),
];

/// Destinations offered in the Intercity vertical.
final List<Place> intercityPlaces = [
  _p('ic_ipoh', 'Ipoh', 'Perak', 4.5975, 101.0901, PlaceCategory.area),
  _p('ic_penang', 'George Town, Penang', 'Pulau Pinang', 5.4141, 100.3288, PlaceCategory.area),
  _p('ic_melaka', 'Melaka', 'Melaka', 2.1896, 102.2501, PlaceCategory.area),
  _p('ic_jb', 'Johor Bahru', 'Johor', 1.4927, 103.7414, PlaceCategory.area),
  _p('ic_seremban', 'Seremban', 'Negeri Sembilan', 2.7297, 101.9381, PlaceCategory.area),
  _p('ic_kuantan', 'Kuantan', 'Pahang', 3.8077, 103.3260, PlaceCategory.area),
  _p('ic_cameron', 'Cameron Highlands', 'Pahang', 4.4711, 101.3776, PlaceCategory.area),
];

final List<Place> allPlaces = [...places, ...intercityPlaces];

/// Street names used to synthesise "dropped pin" addresses.
const streets = [
  'Jalan Ampang', 'Jalan Sultan Ismail', 'Jalan Raja Chulan', 'Jalan Tun Razak',
  'Jalan Bukit Bintang', 'Jalan Imbi', 'Jalan Pudu', 'Jalan Kuching',
  'Jalan Ipoh', 'Jalan Cheras', 'Jalan Klang Lama', 'Jalan Duta',
  'Persiaran KLCC', 'Lebuh Ampang', 'Jalan Sultan Hishamuddin', 'Jalan Damansara',
];

List<Place> fuzzySearch(String query, {List<Place>? pool, int limit = 8}) {
  final q = query.trim().toLowerCase();
  if (q.isEmpty) return const [];
  final source = pool ?? allPlaces;
  final words = q.split(RegExp(r'\s+'));

  final scored = <({Place place, int score})>[];
  for (final place in source) {
    final name = place.name.toLowerCase();
    final address = place.address.toLowerCase();
    var score = 0;
    if (name.startsWith(q)) {
      score += 100;
    } else if (name.contains(q)) {
      score += 60;
    }
    if (address.contains(q)) score += 25;
    // Match on individual words so "mid valley" finds "Mid Valley Megamall".
    if (words.length > 1 &&
        words.every((w) => name.contains(w) || address.contains(w))) {
      score += 40;
    }
    if (score > 0) scored.add((place: place, score: score));
  }
  scored.sort((a, b) => b.score.compareTo(a.score));
  return scored.take(limit).map((s) => s.place).toList();
}
